---
title: "maftools : Summarize, Analyze and Visualize MAF Files"
author: "Anand Mayakonda"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: true
    self_contained: yes
---


```{r setup}
knitr::opts_chunk$set(error = TRUE)
```

# Installation

```{r}
library(maftools)
library(flexdashboard)
library(tidyverse)
library(ggplot2)
library(readr)
library(maps)
library(plotly)
library(mclust)
library(here)
library(NMF)
library(pheatmap)
```

```{r}
cohort_maf   <- maftools::merge_mafs(paste0(getwd(), "/" , "mafs/",  list.files("mafs/")))
maftools::inferHeterogeneity(cohort_maf)
```



## Mutational Signatures
Every cancer, as it progresses leaves a signature characterized by specific pattern of nucleotide substitutions. [Alexandrov et.al](http://www.nature.com/nature/journal/v500/n7463/full/nature12477.html) have shown such mutational signatures, derived from over 7000 cancer samples [Leiserson et al.,2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4531541/). Such signatures can be extracted by decomposing matrix of nucleotide substitutions, classified into 96 substitution classes based on immediate bases surrounding the mutated base. Extracted signatures can also be compared to those [validated signatures](http://cancer.sanger.ac.uk/cosmic/signatures).

First step in signature analysis is to obtain the adjacent bases surrounding the mutated base and form a mutation matrix. 
NOTE: Earlier versions of maftools required a fasta file as an input. But starting from 1.8.0, BSgenome objects are used for faster sequence extraction.

```{r, eval=TRUE}
#Requires BSgenome object
library(BSgenome.Hsapiens.UCSC.hg19, quietly = TRUE)
trinucleotide_matrix = trinucleotideMatrix(maf        = cohort_maf,
                                           prefix     = '',
                                           add        = TRUE,
                                           ref_genome = "BSgenome.Hsapiens.UCSC.hg19")
```

trinucleotideMatrix {maftools}	R Documentation
Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores.
Description
Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores.

Usage
trinucleotideMatrix(maf, ref_genome = NULL, prefix = NULL,
  add = TRUE, ignoreChr = NULL, useSyn = TRUE, fn = NULL)
Arguments
maf	
an MAF object generated by read.maf

ref_genome	
BSgenome object or name of the installed BSgenome package. Example: BSgenome.Hsapiens.UCSC.hg19 Default NULL, tries to auto-detect from installed genomes.

prefix	
Prefix to add or remove from contig names in MAF file.

add	
If prefix is used, default is to add prefix to contig names in MAF file. If false prefix will be removed from contig names.

ignoreChr	
Chromsomes to ignore from analysis. e.g. chrM

useSyn	
Logical. Whether to include synonymous variants in analysis. Defaults to TRUE

fn	
If given writes APOBEC results to an output file with basename fn. Default NULL.

Details
Extracts immediate 5' and 3' bases flanking the mutated site and classifies them into 96 substitution classes. Requires BSgenome data packages for sequence extraction.

APOBEC Enrichment: Enrichment score is calculated using the same method described by Roberts et al.

E = (n_tcw * background_c) / (n_C * background_tcw)

where, n_tcw = number of mutations within T[C>T]W and T[C>G]W context. (W -> A or T)

n_C = number of mutated C and G

background_C and background_tcw motifs are number of C and TCW motifs occuring around +/- 20bp of each mutation.

One-sided Fisher's Exact test is performed to determine the enrichment of APOBEC tcw mutations over background.

Value
list of 2. A matrix of dimension nx96, where n is the number of samples in the MAF and a table describing APOBEC enrichment per sample.

>References
Roberts SA, Lawrence MS, Klimczak LJ, et al. An APOBEC Cytidine Deaminase Mutagenesis Pattern is Widespread in Human Cancers. Nature genetics. 2013;45(9):970-976. doi:10.1038/ng.2702.


```{r}
head(trinucleotide_matrix$nmf_matrix)
```

```{r}
head(trinucleotide_matrix$APOBEC_scores)
```


Above function performs two steps:

  * Prepares a mutational matrix for signature analysis.
  * Estimates APOBEC enrichment scores

### APOBEC Enrichment estimation.
APOBEC induced mutations are more frequent in solid tumors and are mainly associated with C>T transition events occurring in TCW motif. APOBEC enrichment scores in the above command are estimated using the method described by [Roberts et al](https://www.nature.com/articles/ng.2702). Briefly, enrichment of C>T mutations occurring within TCW motif over all of the C>T mutations in a given sample is compared to background Cytosines and TCWs occurring within 20bp of mutated bases.


$$\frac{n_{tCw} * background_C}{n_C * background_{TCW}}$$

One-sided fishers exact test is also performed to statistically evaluate the enrichment score, as described in original study by Roberts et al.


### Differences between APOBEC enriched and non-enriched samples

We can also analyze the differences in mutational patterns between APOBEC enriched and non-APOBEC enriched samples. `plotApobecDiff` is a function which takes APOBEC enrichment scores estimated by `trinucleotideMatrix` and classifies samples into APOBEC enriched and non-APOBEC enriched. Once stratified, it compares these two groups to identify differentially altered genes.

```{r, eval=TRUE, fig.height=4, fig.width=7}
plotApobecDiff(tnm = trinucleotide_matrix,
               maf = cohort_maf,
               pVal = 0.2)
```


### Signature analysis
`extractSignatures` uses non-negative matrix factorization to decompose nx96 dimension matrix into r signatures, where n is the number of samples from input MAF [11](#references). By default function runs nmf on 6 ranks and chooses the best possible value based on maximum cophenetic-correlation coefficient. It is also possible to manually specify r. Once decomposed, signatures are compared against known signatures derived from [Alexandrov et.al](http://www.nature.com/nature/journal/v500/n7463/full/nature12477.html), and cosine similarity is calculated to identify best match.


```{r, fig.height=5, fig.width=5, eval=FALSE, message=FALSE}
#Run main function with maximum 6 signatures. 
library('NMF')
mutsigs_nTry6 = extractSignatures(mat            = trinucleotide_matrix,
                                  nTry           = 6,
                                  plotBestFitRes = TRUE)
```

```{r, fig.width=6, fig.height=4, fig.align='center', eval = T}
plotSignatures(mutsigs_nTry6,
               title_size = 0.8, )
```

```{r, fig.height=5, fig.width=5, eval=TRUE, message=FALSE, echo=FALSE}
#Run main function with maximum 6 signatures.
library('NMF')
mutsigs_n2 = extractSignatures(mat            = trinucleotide_matrix,
                               n              = 2,
                               plotBestFitRes = FALSE,
                               pConstant      = 0.1)
```

```{r, fig.width=6, fig.height=4, fig.align='center', eval = T}
plotSignatures(mutsigs_n2,
               title_size = 0.8, )
```

`extractSignatures` gives a warning that no mutations are found for class A[T>G]C conversions. This is possible when the number of samples are low or in tumors with low mutation rate, such as in the case of Leukemia. In this scenario, a small positive value is added to avoid computational difficulties. It also prints other statistics for range of values that were tried, and chooses the rank with highest cophenetic metric. Above stats should give an estimate of range of best possible r values and in case the chosen r is overfitted/underfitted, it is also possible to be re-run `extractSignatures` by manually specifying r.

Once decomposed, signatures are compared against known and validated signatures from [Sanger](https://www.nature.com/articles/leu201669). See [here](http://cancer.sanger.ac.uk/cosmic/signatures) for list of validated signatures. In the above example, 2 signatures are derived, which are similar to validated Signature-1 and Signature-5. Signature_1 is a result of elevated rate of spontaneous deamination of 5-methyl-cytosine, resulting in C>T transitions and predominantly occurs at NpCpG trinucleotide, which is a most common process in AML [link](https://www.sciencedirect.com/science/article/abs/pii/S1368837512002977?via%3Dihub).

# Cosine similarities against validated signatures
Full table of cosine similarities against validated signatures are also returned, which can be further analysed. Below plot shows comparison of similarities of detected signatures against validated signatures.

```{r, fig.width=7, fig.height=2.5, fig.align='center'}
library('pheatmap')
pheatmap::pheatmap(mat          = mutsigs_nTry6$coSineSimMat,
                   cluster_rows = FALSE, 
                   main         = "cosine similarity against validated signatures"
                   )
```


### Signature enrichment analysis
Signatures can further be assigned to samples and enrichment analysis can be performd using `signatureEnrichment` funtion, which identifies mutations enriched in every signature identified.

```{r, echo=FALSE,eval=FALSE}
colnames(mutsigs_nTry6$contributions) = as.character(getSampleSummary(x = cohort_maf)[,Tumor_Sample_Barcode])
```

```{r}
signature_enrichment = signatureEnrichment(maf     = cohort_maf,
                                           sig_res = mutsigs_nTry6)
```

```{r}
head(signature_enrichment$pairwise_comparision)
```

```{r}
head(signature_enrichment$groupwise_comparision)
```

```{r}
head(signature_enrichment$cf_sizes)
```

```{r}
head(signature_enrichment$clinicalFeature)
```

```{r}
head(signature_enrichment$Signature_Assignment)
```

```{r}
head(signature_enrichment$mutation_load)
```



```{r, fig.height=3.5, fig.width=5, warning=FALSE}
signature_enrichment = signatureEnrichment(maf     = cohort_maf,
                                           sig_res = mutsigs_nTry6,
                                           minMut  = 5 ,  #default
                                           )
```

Above results can be visualized similar to clinical enrichments.

# plotEnrichmentResults  pVal = 0.05
```
```{r, fig.height=4, fig.width=6}
plotEnrichmentResults(enrich_res = signature_enrichment,
                      pVal = 0.05)
```

# plotEnrichmentResults  pVal = 0.1
```
```{r, fig.height=4, fig.width=6}
plotEnrichmentResults(enrich_res = signature_enrichment,
                      pVal = 0.1, 
                      showTitle = TRUE)
```